feat: add proposal for dispose functionality (#46)
open-feature/ofep: 009-OFEP-add-dispose.md

## Add dispose functionality to API

## State: DRAFTING

> The goal of this OFEP is to enhance the OpenFeature API to allow authors to cleanly dispose of any resources by consumed by OpenFeature providers.

Reviewer: ```suggestion
The goal of this OFEP is to enhance the OpenFeature API to allow authors to cleanly dispose of any resources consumed by OpenFeature providers.
```

## Background

When implementing OpenFeature providers, you sometimes need to register event handlers, or timers or similar functionality to
correctly handle functionality of the vendor library. Some vendor libraries might also have analytics, or debugging functionality
that can be enabled and might get flushed to the vendor platform when the library gets stopped or disposed off.

Currently, you aren't able to cleanly handle these use cases in OpenFeature providers.

## Proposal

To allow to cleanly dispose of resources consumed by OpenFeature providers, I would like to purpose a `dispose`-function that
can be called by OpenFeature SDK to allow to cleanup these resources. The OpenFeature provider will implement an async function

> named `dispose` that can be called by the internal OpenFeature SDK.

Reviewer 2: Should we have a comment here for idiomatic alternatives? "`dispose` or technology idiom equivalent"? However similar things have been worded in the spec.

For instance, in .Net we should likely do this by implementing `IDisposable` (which happens to be the same name for the method).  

Also, being async would probably not be appropriate for some languages. So maybe `async` when reasonable?

Reviewer 3: Yeah, I think it's reasonable to write down something like what the best practice is for the chosen programming language? E.g. `IDisposable` in .Net, `Destroy` called by `Free` in Delphi etc.

Author: > Should we have a comment here for idiomatic alternatives? "dispose or technology idiom equivalent"? However similar things have been worded in the spec.

I'll add wording like you describe here as well.

> Also, being async would probably not be appropriate for some languages. So maybe async when reasonable?

Agreed.

Author: I've made this update.

 
### Example Implementation

> The `Provider` type gets extended with the following function:

Author: ```suggestion
The `Provider` interfaces gets extended (in a non-breaking way) to include the following function:
```

## Add dispose functionality to API

## State: DRAFTING

The goal of this OFEP is to enhance the OpenFeature API to allow authors to cleanly dispose of any resources consumed by OpenFeature providers.

## Background

When implementing OpenFeature providers, you sometimes need to register event handlers, or timers or similar functionality to
correctly handle functionality of the vendor library. Some vendor libraries might also have analytics, or debugging functionality
that can be enabled and might get flushed to the vendor platform when the library gets stopped or disposed off.

Currently, you aren't able to cleanly handle these use cases in OpenFeature providers.

## Proposal

> To allow clean dispose of resources consumed by OpenFeature providers, I would like to propose a function that can be called by OpenFeature SDK that allow to cleanup these resources.

Author: ```suggestion
To allow clean disposal of resources consumed by OpenFeature providers, I would like to propose a function that can be called by OpenFeature SDK that allow to cleanup these resources.
```

> To allow to cleanly dispose of resources consumed by OpenFeature providers, I would like to purpose a function that can be called by OpenFeature SDK to allow to cleanup these resources.

Reviewer 4: ```suggestion
To allow clean dispose of resources consumed by OpenFeature providers, I would like to propose a function that can be called by OpenFeature SDK that allow to cleanup these resources.
```

## Add dispose functionality to API

> ## State: DRAFTING

Reviewer: Reminder to update the state once a consensus has been made.

The goal of this OFEP is to enhance the OpenFeature API to allow authors to cleanly dispose of any resources by consumed by OpenFeature providers.

## Background

When implementing OpenFeature providers, you sometimes need to register event handlers, or timers or similar functionality to
correctly handle functionality of the vendor library. Some vendor libraries might also have analytics, or debugging functionality
that can be enabled and might get flushed to the vendor platform when the library gets stopped or disposed off.

Currently, you aren't able to cleanly handle these use cases in OpenFeature providers.

## Proposal

To allow to cleanly dispose of resources consumed by OpenFeature providers, I would like to purpose a `dispose`-function that
can be called by OpenFeature SDK to allow to cleanup these resources. The OpenFeature provider will implement an async function

> named `dispose` that can be called by the internal OpenFeature SDK

Reviewer: ```suggestion
named `dispose` that can be called by the internal OpenFeature SDK.
```

> ### [Example] Implementation

Reviewer: Why is example wrapping in brackets?

Author: ```suggestion
### Example Implementation
```

Author: I think it was just copied from the template.

> The `Provider` type gets extended with the following function:

Reviewer: I think this should be optional on the provider interface to avoid introducing a breaking change.

Author: Yep, agreed. That's why I used a new interface in my example. I will make it clearer in the OFEP though.

Author: Done. I was intentionally vague here since there's a number of ways to do this sort of thing and I think that should be left up to the SDK authors.

```typescript
dispose(): Promise<void>
```

> The function can be called by the the `API` instance when it requests the resources need to be disposed off, for this reason, 

Reviewer: ```suggestion
The function can be called by the the `API` instance when it requests the resources that need to be disposed off, for this reason, 
```

the global API needs to be enhanced to also have the `dispose`-function:

```typescript
dispose(): Promise<void>
```

A potential implementation could be the following in the Node SDK:

```typescript

> /**
>  * Dispose
>  */

Reviewer: ```suggestion

```
